luigi/contrib/pai.py

```diff
@@ -0,0 +1,311 @@
+        default='admin',
+        description='your username')
+    password = luigi.Parameter(
+        default=None,
+        description='your password')
+    expiration = luigi.IntParameter(
+        default=3600,
+        description='expiration time in seconds')
+
+
+class PaiTask(luigi.Task):
+    __POLL_TIME = 5
+
+    @abc.abstractproperty
+    def name(self):
+        """Name for the job, need to be unique, required"""
+        return 'SklearnExample'
+
+    @abc.abstractproperty
+    def image(self):
+        """URL pointing to the Docker image for all tasks in the job, required"""
+        return 'openpai/pai.example.sklearn'
+
+    @abc.abstractproperty
+    def tasks(self):
+        """List of taskRole, one task role at least, required"""
+        return []
+
+    @property
+    def auth_file_path(self):
+        """Docker registry authentication file existing on HDFS, optional"""
+        return None
+
+    @property
+    def data_dir(self):
+        """Data directory existing on HDFS, optional"""
+        return None
+
+    @property
+    def code_dir(self):
+        """Code directory existing on HDFS, should not contain any data and should be less than 200MB, optional"""
+        return None
+
+    @property
+    def output_dir(self):
+        """Output directory on HDFS, $PAI_DEFAULT_FS_URI/$jobName/output will be used if not specified, optional"""
+        return '$PAI_DEFAULT_FS_URI/{0}/output'.format(self.name)
+
+    @property
+    def virtual_cluster(self):
+        """The virtual cluster job runs on. If omitted, the job will run on default virtual cluster, optional"""
+        return 'default'
+
+    @property
+    def gpu_type(self):
+        """Specify the GPU type to be used in the tasks. If omitted, the job will run on any gpu type, optional"""
+        return None
+
+    @property
+    def retry_count(self):
+        """Job retry count, no less than 0, optional"""
+        return 0
+
+    def __init_token(self):
+        self.__openpai = OpenPai()
+
+        request_json = json.dumps({'username': self.__openpai.username, 'password': self.__openpai.password,
+                                   'expiration': self.__openpai.expiration})
+        logger.debug('Get token request {0}'.format(request_json))
+        response = rs.post(urljoin(self.__openpai.pai_url, '/api/v1/token'),
+                           headers={'Content-Type': 'application/json'}, data=request_json)
+        logger.debug('Get token response {0}'.format(response.text))
+        if response.status_code != 200:
+            msg = 'Get token request failed, response is {}'.format(response.text)
+            logger.error(msg)
+            raise Exception(msg)
+        else:
+            self.__token = response.json()['token']
+
+    def __init__(self, *args, **kwargs):
+        """
+        :param pai_url: The rest server url of PAI clusters, default is 'http://127.0.0.1:9186'.
+        :param token: The toke used to auth the rest server of PAI.
+
```

Comment:
remove this unnecessary blank line, please

Reply:
Done